### PR TITLE
implement self-request-focus and demands_attention for xwayland

### PR DESCRIPTION
**Pull request by damianatorrpm**

### CODE REVIEW

**Comment by damianatorrpm** [comment]
Unknown to me why the 3rd check fails, locally git ls-files | grep "hpp$\|cpp$" | xargs uncrustify -c uncrustify.ini --check
succeeds with uncrustify 0.71

File: src/view/xwayland.cpp [hunk 1]
```diff
@@ -326,6 +327,15 @@ class wayfire_xwayland_view : public wayfire_xwayland_view_base
 
         on_request_move.set_callback([&] (void*) { move_request(); });
         on_request_resize.set_callback([&] (void*) { resize_request(); });
+        on_request_activate.set_callback([&] (void*)
+        {
+            if (!this->activated) {
```
Comment (ammen99) [inline src/view/xwayland.cpp:332]: not sure if this will fix it, but `{` should be on the next line anyway.

Comment (damianatorrpm) [reply src/view/xwayland.cpp]: yes, that was the problem it seems

File: src/view/xwayland.cpp [hunk 2]
```diff
@@ -326,6 +327,16 @@ class wayfire_xwayland_view : public wayfire_xwayland_view_base
 
         on_request_move.set_callback([&] (void*) { move_request(); });
         on_request_resize.set_callback([&] (void*) { resize_request(); });
+        on_request_activate.set_callback([&] (void*)
+        {
+            if (!this->activated)
+            {
+                view_self_request_focus_signal data;
+                data.view = this;
+                this->get_output()->emit_signal("view-self-request-focus", &data);
```
Comment (ammen99) [inline src/view/xwayland.cpp:336]: In theory it is possible to have `this->get_output()` return `NULL`. The codebase is not really ready for that but we should check it for new code (also same thing for the other signal). Aside from that LGTM.

File: src/api/wayfire/signal-definitions.hpp [hunk 3]
```diff
@@ -75,7 +75,18 @@ struct view_tiled_signal : public _view_signal
  * view should become active.
  */
 struct view_self_request_focus_signal : public _view_signal
-{};
+{
+    bool wants_focus = true;
```
Comment (ammen99) [inline src/api/wayfire/signal-definitions.hpp:79]: looks like this is unused?

**Comment by damianatorrpm** [comment]
@ammen99 
running uncrustify locally gives me this:
```
@@ -332,19 +332,18 @@ class wayfire_xwayland_view : public wayfire_xwayland_view_base
             if (!this->activated)
             {
                 view_self_request_focus_signal data;
                 wf::output_t *view_output;
                 view_output = this->get_output();
-                data.view = this;
+                data.view   = this;
                 if (view_output)
                 {
                     view_output->emit_signal("view-self-request-focus", &data);
-                }
-                else
+                } else
                 {
                     wf::get_core().emit_signal("view-self-request-focus", &data);
-                    this->emit_signal("view-self-request-focus", &data); 
+                    this->emit_signal("view-self-request-focus", &data);
                 }
             }
         });
 
         on_request_maximize.set_callback([&] (void*)
@@ -378,20 +377,20 @@ class wayfire_xwayland_view : public wayfire_xwayland_view_base
         on_set_hints.set_callback([&] (void*)
         {
             view_hints_changed_signal data;
             wf::output_t *view_output;
             view_output = this->get_output();
-            data.view = this;
+            data.view   = this;
             if (xw->hints_urgency)
             {
                 data.demands_attention = true;
             }
+
             if (view_output)
             {
                 view_output->emit_signal("view-hints-changed", &data);
-            }
-            else
+            } else
             {
                 wf::get_core().emit_signal("view-hints-changed", &data);
                 this->emit_signal("view-hints-changed", &data);
             }
         });
```

File: src/view/xwayland.cpp [hunk 4]
```diff
@@ -326,6 +327,25 @@ class wayfire_xwayland_view : public wayfire_xwayland_view_base
...
         on_request_resize.set_callback([&] (void*) { resize_request(); });
+        on_request_activate.set_callback([&] (void*)
+        {
+            if (!this->activated)
+            {
+                view_self_request_focus_signal data;
+                wf::output_t *view_output;
+                view_output = this->get_output();
+                data.view   = this;
+                if (view_output)
```
Comment (ammen99) [inline src/view/xwayland.cpp:338]: I thought we discussed this a second time on IRC, and this signal should be emitted on the view and on core always?

File: src/view/xwayland.cpp [hunk 5]
```diff
@@ -354,11 +374,34 @@ class wayfire_xwayland_view : public wayfire_xwayland_view_base
...
+            data.view   = this;
+            if (xw->hints_urgency)
+            {
+                data.demands_attention = true;
+            }
+
+            if (view_output)
+            {
+                view_output->emit_signal("view-hints-changed", &data);
+            } else
```
Comment (ammen99) [inline src/view/xwayland.cpp:391]: and same here

Comment (damianatorrpm) [reply src/view/xwayland.cpp]: done

File: src/view/xwayland.cpp (hunk 4 above)
Comment (damianatorrpm) [reply src/view/xwayland.cpp]: done

**Comment by damianatorrpm** [comment]
is it possible that the checks don't run because of above 'conflict'?
I don't know how to fix that.

**Comment by ammen99** [comment]
> is it possible that the checks don't run because of above 'conflict'?
> I don't know how to fix that.

You can just try to rebase on master and git will tell you where the conflicts are.

**Comment by damianatorrpm** [comment]
> > is it possible that the checks don't run because of above 'conflict'?
> > I don't know how to fix that.
> 
> You can just try to rebase on master and git will tell you where the conflicts are.

I am lost here.

If I follow the instructions to update https://github.com/damianatorrpm/wayfire/
to master, it will have another commit used to update it from WayfireWM/wayfire
which also appears in the pull request.
https://rick.cogley.info/post/update-your-forked-repository-directly-on-github/#:~:text=Access%20your%20forked%20repository%20on,original%20to%20the%20head%20fork.

Same goes for damianatorrpm:xwayland_view_attention.
No idea how to do that.

**Comment by kode54** [comment]
```
git checkout master
git pull origin master # this doesn't need to rebase, since you're not working in master
git checkout xwayland_view_attention
git rebase master
```

Follow the prompts from the rebase to modify your commits to succeed, then `git add` the files you've fixed, and `git rebase continue` to flow to the next commit, or else complete the rebase.

**Comment by damianatorrpm** [comment]
> ```
> git rebase master
> ```

```
tux@localhost:~/wayfire_git> git checkout master
Switched to branch 'master'
tux@localhost:~/wayfire_git> git pull origin master
warning: Pulling without specifying how to reconcile divergent branches is
discouraged. You can squelch this message by running one of the following
commands sometime before your next pull:

  git config pull.rebase false  # merge (the default strategy)
  git config pull.rebase true   # rebase
  git config pull.ff only       # fast-forward only

You can replace "git config" with "git config --global" to set a default
preference for all repositories. You can also pass --rebase, --no-rebase,
or --ff-only on the command line to override the configured default per
invocation.

From https://github.com/damianatorrpm/wayfire
 * branch            master     -> FETCH_HEAD
Already up to date.
tux@localhost:~/wayfire_git> git checkout xwayland_view_attention
Switched to branch 'xwayland_view_attention'
Your branch is up to date with 'origin/xwayland_view_attention'.
tux@localhost:~/wayfire_git> git rebase master
Current branch xwayland_view_attention is up to date.
```
 But the master @ https://github.com/damianatorrpm/wayfire 
 is not up-to-date with https://github.com/WayfireWM/wayfire
...

**Comment by ammen99** [comment]
@damianatorrpm You can add a new remote in your local repository:

```
git remote add upstream https://github.com/WayfireWM/wayfire
```

Then you can use the upstream's master branch to rebase:

```
git fetch upstream
git rebase upstream/master
```

**Pull request closed without merging**